android/src/main/java/com/mattermost/networkclient/helpers/UploadProgressListener.kt

```diff
@@ -0,0 +1,30 @@
+package com.mattermost.networkclient.helpers
+
+import com.facebook.react.bridge.Arguments
+import com.facebook.react.bridge.ReactContext
+import com.facebook.react.modules.core.DeviceEventManagerModule.RCTDeviceEventEmitter
+import com.mattermost.networkclient.enums.APIClientEvents
+import java.math.RoundingMode
+import java.text.DecimalFormat
+
+interface ProgressListenerInterface {
+    fun update(bytesRead: Double, contentLength: Double)
+    fun emitProgressEvent(progress: Double)
+}
+
+class ProgressListener(private val reactContext: ReactContext, private val taskId: String) : ProgressListenerInterface {
+    // 2-Decimal places, rounded up
+    private val df = DecimalFormat("#.00", ).apply{ roundingMode = RoundingMode.UP }
+
+    override fun emitProgressEvent(progress: Double) {
+        val params = Arguments.createMap()
+        params.putDouble("fractionCompleted", progress)
+        params.putString("taskId", "$taskId")
+        reactContext.getJSModule(RCTDeviceEventEmitter::class.java).emit(APIClientEvents.UPLOAD_PROGRESS.event, params)
+    }
+
+    override fun update(bytesRead: Double, contentLength: Double) {
+        // Only emit if we can show progress against content length
+        if(contentLength > 0) emitProgressEvent(df.format(bytesRead / contentLength).toDouble())
```

Comment:
@enahum / @migbot - We would still want to show some progress on the upload bar no? I'm wondering if there's ever a scenario where it's unable to compute the length of the stream but a stream still exists that it starts uploading...

Reply:
If no content lenght I would return the bytesRead..